### PR TITLE
Weights and Biases

### PR DESCRIPTION
**Pull request by dylanuys**
- Validator logging to wandb includes
    - Bitmind codebase version and validator uid (these will be present in the run name rather than individual logs)
    - Image label (1/0)
    - Miner UIDs and predictions
    - If label == 1
        - the generated image
        - the associated prompt
        - prompt generation technique/version/model (TODO - decide how to version this once new prompt gen pipeline is done)
        - diffuser name
    - if label == 0
        - the id/index of the image (this would change with the swapping of HF for IPFS/other storage)
        - dataset name

### CODE REVIEW

File: bitmind/validator/forward.py [hunk 1]
```diff
@@ -103,12 +103,38 @@ async def forward(self):
     )
 
     rewards = get_rewards(label=label, responses=responses)
+
```
Comment (benliang99) [inline bitmind/validator/forward.py:98]: Question: Do we want to have both CSV logging and optional wandb logging? Also, do we still want to track miner responses?

Added some comments and reorganized code.

**Review by benliang99 — COMMENTED** [review]
Looks good, will approve when comments are addressed!

File: bitmind/validator/forward.py (hunk 1 above)
Comment (dylanuys) [reply bitmind/validator/forward.py:98]: Nope just wandb, will remove csv code now

Comment (dylanuys) [reply bitmind/validator/forward.py:98]: when you say track miner performance do you mean the print statements? miner responses and image labels are logged to wandb, from which we can infer miner performance.

Not sure whether it's better to precompute correctness on validators and log that, or build reporting on top of wandb to compute it from predicitons/labels. I'm leaning towards the latter, but open to input!

Comment (benliang99) [reply bitmind/validator/forward.py:98]: Yes, I meant the print statements that were there.
`   self.last_responding_miner_uids = []
    for i, pred in enumerate(responses):
        # Logging specific prediction details
        if pred != -1:
            bt.logging.info(f'Miner uid: {miner_uids[i]} | prediction: {pred} | correct: {np.round(pred) == label} | reward: {rewards[i]}')
            self.last_responding_miner_uids.append(miner_uids[i])`

I vote for consolidating all logging efforts where possible in wandb for simplicity.

Comment (dylanuys) [reply bitmind/validator/forward.py:98]: cool, agreed. I'm going to leave this logging statement in place for now - we may eventually decide to remove it if we deem it redundant, but for now it's helpful to see what miners are responding in realtime when looking at validator logs. Sound ok?

**Review by benliang99 — APPROVED** [review]

**Pull request merged**